src/learning/InitializeProfile.cpp

```diff
@@ -0,0 +1,203 @@
+}
+
+std::vector<float> ProfileInitializer::categoryFrequency() {
+
+  // Extraction category rank from each alternative and storing it in values
+  std::unordered_map<std::string, Category> map =
+      altPerformance_.getAlternativesAssignments();
+  std::vector<int> values(map.size());
+
+  auto cat_value_selector = [](auto pair) {
+    return pair.second.getCategoryRank();
+  };
+
+  std::transform(map.begin(), map.end(), values.begin(), cat_value_selector);
+
+  // Counting the number of occurence for each category rank
+  std::unordered_map<int, int> counts;
+  for (auto v : values)
+    ++counts[v];
+
+  // Counting the number of alternatives in the altPerformance object
+  const int nbAlternatives = std::accumulate(
+      std::begin(counts), std::end(counts), 0,
+      [](int value, const std::unordered_map<int, int>::value_type &p) {
+        return value + p.second;
+      });
+
+  // Computing the frequency
+  std::vector<float> frequency;
+  for (auto const pair : counts) {
+    float c = pair.second;
+    frequency.push_back(c / nbAlternatives);
+  }
+  return frequency;
+}
+
+std::vector<std::string> ProfileInitializer::getProfilePerformanceCandidates(
+    const Criterion &crit, const Category &cat, const int nbCategories) {
+
+  int catAbove;
+  int catBelow;
+  std::vector<std::string> candidates;
+
+  // Category rank starts at 0
+  if (cat.getCategoryRank() == nbCategories - 1) {
+    catAbove = cat.getCategoryRank();
+    catBelow = catAbove - 1;
+  } else {
+    catBelow = cat.getCategoryRank();
+    catAbove = catBelow + 1;
+  }
+
+  for (auto pair : altPerformance_.getAlternativesAssignments()) {
+    if (pair.second.getCategoryRank() == catBelow ||
+        pair.second.getCategoryRank() == catAbove) {
+      candidates.push_back(pair.first);
+    }
+  }
+  return candidates;
+}
+
+float ProfileInitializer::weightedProbability(
+    const std::string altId, const Criterion &crit, const Category &catAbove,
+    const Category &catBelow, const int nbCategories,
+    const std::vector<float> &catFrequency, std::vector<std::string> candidates,
+    float delta) {
+  // creating imaginary profile performance for criterion crit
+  std::string critId = crit.getId();
+  float imaginaryProfilePerformance =
+      altPerformance_.getPerf(altId, critId).getValue() + delta;
+
+  // Creating 2 int that will count the number of correctly classified
+  // alternatives for criterion crit for a profile performance of
+  // imaginaryProfilePerformance.
+  int catAboveCounter = 0;
+  int catBelowCounter = 0;
+
+  // candidates are computed thanks to the method
+  // ProfileInitializer::getProfilePerformanceCandidates(crit, catBelow,
+  // nbCategories);
+  for (std::string can : candidates) {
+    // if the performance of the candidate is higher than the profile
+    // performance and that the alternatives category is catAbove then we have
+    // correctly classified it
+    if (altPerformance_.getPerf(can, critId).getValue() >
```

Comment:
getPerf here and in L130 should be avoided as well, I see you might need to give the whole row of crit to this function instead of a single perf.

Reply:
Dont get it

Reply:
need to not use [] operator since it does a search in a n**2 matrix and is slow. Better use to go through PT once and get std::Vector of Perf

Reply:
done